### DISCUSSION TITLE
Why array-key is the supertype but not a union of int and string

### DISCUSSION THREAD

**Discussion by vudaltsov** [vimeo/psalm#9403] [General]
https://psalm.dev/docs/annotating_code/type_syntax/scalar_types/#array-key

I am just curious, why `array-key` is defined as supertype but not union? What is the difference in this particular case?

**Comment by vudaltsov** [comment]
Same question for scalar: https://psalm.dev/docs/annotating_code/type_syntax/scalar_types/#scalar.

**Comment by orklah** [comment]
The array-key usefulness is best described with this snippet: https://psalm.dev/r/1a72bb6061

the severity of the issue is not the same when it's an array-key vs string|int

scalar may be the same even if i don't see the need for that. It may just be a better way to write string|int|float|bool and that's it

**Reply by vudaltsov** [reply]
Thank you, that makes sense for `array-key`!

**Reply by vudaltsov** [reply]
So, basically, this only affects the error level, right? And that might be crucial for projects with high errorLevels?

**Reply by orklah** [reply]
crucial is a strong word, let's say it allows to focus on more important issues first.

Though to be honest, I prefer PHPStan way of handling that sort of thing. They have a BenevolentUnionType instead of a UnionType that will allow Coercion into a subtype if necessary. It's much more flexible than just having a special type for a special case
